File: solutions/elixir/resistor-color-trio/1/lib/resistor_color_trio.ex
defmodule ResistorColorTrio do
  @colors %{
    black: 0,
    brown: 1,
    red: 2,
    orange: 3,
    yellow: 4,
    green: 5,
    blue: 6,
    violet: 7,
    grey: 8,
    white: 9
  }

  @doc """
  Calculate the resistance value in ohm or kiloohm from resistor colors
  """
  @spec label(colors :: [atom]) :: {number, :ohms | :kiloohms}
  def label([decimal, units, zeros | _]) do
    val = (@colors[decimal] * 10 + @colors[units]) * 10 ** @colors[zeros]

    if val > 1000,
      do: {val / 1000, :kiloohms},
      else: {val, :ohms}
  end
end

File: solutions/ruby/resistor-color-trio/1/resistor_color_trio.rb
# frozen_string_literal: true

# ResistorColorTrio class
class ResistorColorTrio
  SCORES = %w[black brown red orange yellow green blue violet grey white].freeze

  def initialize(colors)
    raise ArgumentError unless (colors - SCORES).size.zero?

    @colors = colors[0..1]
    @zeroes = SCORES.index(colors[2])
  end

  def label
    kilo = score > 1000 ? 'kilo' : ''
    value = score > 1000 ? score / 1000 : score
    "Resistor value: #{value} #{kilo}ohms"
  end

  def score
    (@colors.map { SCORES.index(_1).to_s } + %w[0] * @zeroes)
      .join.to_i
  end
end
